### PR TITLE
Implements project environments #164

### PR DESCRIPTION
**Pull request by ss7548**

### CODE REVIEW

**Comment by gdubicki** [comment]
Hey @ss7548 , that’s a great looking PR, thanks!

I may have one or two nit-picking comments but I think this will land in main within a few days and gets release soon after. :)

Before I go into the code deeper, can you please just check why the tests failed?

**Comment by ss7548** [comment]
@gdubicki , I guess calling a a super function overwrote or otherwise corrupted memory impacting one of the merge request tests.

File: gitlabform/gitlab/environments.py
Comment (gdubicki) [inline gitlabform/gitlab/environments.py:55]: Is there a specific reason this is not implemented? According to the API docs at https://docs.gitlab.com/ee/api/environments.html#edit-an-existing-environment you can edit an existing environment.

File: gitlabform/gitlabform/processors/project/environments_processor.py
Comment (gdubicki) [inline gitlabform/gitlabform/processors/project/environments_processor.py:19]: Nit: single-letter variables are not great except maybe iterators in loops as they are very uninformative and can be misleading. When I was reading the code below guessed that `e` is a single environment and I was wrong...

Comment (gdubicki) [reply gitlabform/gitlabform/processors/project/environments_processor.py:19]: Also when debug output will be enabled then this line will do 2nd call to get all the environments to GitLab in a row. Check out https://github.com/egnyte/gitlabform/blob/v2/gitlabform/gitlabform/processors/project/project_push_rules_processor.py#L17-L19 where we do it only once.

File: gitlabform/gitlabform/processors/project/environments_processor.py
Comment (gdubicki) [inline gitlabform/gitlabform/processors/project/environments_processor.py:25]: I like that you are getting all the envs and then operate only the ones that you should - that minimizes the number of calls to GitLab. 🙂 It also opens room to easily add "enforce" option - that would keep the envs exactly as defined in the code, so deleting all the ones that may have been defined manually - in the future. 👍

**Review by gdubicki — CHANGES_REQUESTED** [review]
Thanks for fixing the tests! Please check out my review.

**Comment by ss7548** [comment]
Observation on testing.  I test against my dev omnibus install, GitLab V 13.9.2-ee
After running tests, several GROUPS and PROJECTS are left in the wake.
When running the tests for the first time, the test TestMergeRequestApprovers::test__if_it_works__single_user fails
When running the tests for the second time, the aforementioned test passes, but the test TestArchiveProject::test__dont_edit_archived_project fails.

After cleaning up (i.e. removing the GROUPS left over from the tests), the same pattern occurs.

There is a problem with the MergeReqestsApprovers feature which I have seen in my production environment.

When a new project is created, there is by default an approval rule "Any eligible user" that requires 0 approvals.

When you add an approval rule with GitLabForm, it changes the 0 approvals of  "Any eligible user" to the number specified in the approval config.

This is also why the "test__if_it_works__single_user" fails the first time, After setting the approval rule, it expects that there is only one, ignoring the "Any eligible user" rule.   After the test it deletes ALL rules, so the second time you run it, the test passes because the "Any eligible user" rule has been deleted.  

The reason for this is that the PROJECT that this test creates, is left in tact when the tests complete.

I have no explanation for why the TestArchiveProject::test__dont_edit_archived_project test fails the second time through.

**Comment by ss7548** [comment]
@gdubicki  any feedback?

**Comment by gdubicki** [comment]
Hey @ss7548 , sorry for lack of response. I have been looking into this but haven't replied yet because this need more time from me and I wasn't able to find it recently. I will try to do it over this weekend.

**Comment by ss7548** [comment]
not sure if you are waiting for something from me?

**Comment by gdubicki** [comment]
Sorry, I didn't have the time and will to do this before, but I have found it today. I am now trying to overhaul the tests in GitLabForm completely - that should fix all the issues you have reported, @ss7548 . Stay tuned, I will share an update today.

**Comment by gdubicki** [comment]
Unfortunately I got stuck because of #196 . :(

**Comment by gdubicki** [comment]
Thanks for your patience @ss7548 !

I think that I got #196 fixed and have done the tests revamp for:
```
test_archive_project.py
test_branches.py
test_deploy_keys.py
test_files.py
```
...in the v2 branch. A final version 2 will be released within days, so it would be best if you would rebase your branch against `v2` and target `v2` with your PR. Can you do that please?

**Comment by gdubicki** [comment]
Closing in favor of #203 .

**Pull request closed without merging**
